### PR TITLE
indexer: object read API with options 1/N

### PR DESCRIPTION
**Pull request by gegaowp**
## Description 

Implement object read API with options, with some TODOs
- figure package and object query serving, should both base on `objects` table
- parsed object and display needs impl of module_cache on indexer
- add CI tests for object read

## Test Plan 

local run agains indexer and FN, verify that other than display and content, other fields are the same

```
curl --location --request POST http://127.0.0.1:3030 \
--header 'Content-Type: application/json' \
--data-raw '{ "jsonrpc":"2.0", "id":1, "method":"sui_getObject", "params":["0x1ace65f54d65a96251b3f46bfa720ab65a7ebe0174caa9fa1dc0d65a56ae7872", [true, true, true, true, true, true, true]]}'
{"jsonrpc":"2.0","result":{"status":"Exists","details":{"objectId":"0x1ace65f54d65a96251b3f46bfa720ab65a7ebe0174caa9fa1dc0d65a56ae7872","version":1,"digest":"89EGgBRsPBnasx87FCVuHJUeFpFNzKndubjhP21Bo53S","type":"0x2::staking_pool::StakedSui","owner":{"AddressOwner":"0x00df836a90d2b145f4fb5ceaf53d112d905b65e9a5d6ee3141a917d39a21cfde"},"previousTransaction":"Fde5EQZvYY59Aj4ruZYvWdtWXUorsqHUJASosCdWAQvZ","storageRebate":0,"bcs":{"dataType":"moveObject","type":"0x2::staking_pool::StakedSui","hasPublicTransfer":false,"version":1,"bcsBytes":"Gs5l9U1lqWJRs/Rr+nIKtlp+vgF0yqn6HcDWWlaueHJij/0OUemm6jLBPCc5oxqPNEtVfTQp4Fezd6nEmbm7EwDfg2qQ0rFF9Ptc6vU9ES2QW2XppdbuMUGpF9OaIc/eAAAAAAAAAAAAgGIXXtFYAAA="}}},"id":1}%                                         
curl --location --request POST http://127.0.0.1:9000 \
--header 'Content-Type: application/json' \
--data-raw '{ "jsonrpc":"2.0", "id":1, "method":"sui_getObject", "params":["0x1ace65f54d65a96251b3f46bfa720ab65a7ebe0174caa9fa1dc0d65a56ae7872", [true, true, true, true, true, true, true]]}'
{"jsonrpc":"2.0","result":{"status":"Exists","details":{"objectId":"0x1ace65f54d65a96251b3f46bfa720ab65a7ebe0174caa9fa1dc0d65a56ae7872","version":1,"digest":"89EGgBRsPBnasx87FCVuHJUeFpFNzKndubjhP21Bo53S","type":"0x2::staking_pool::StakedSui","owner":{"AddressOwner":"0x00df836a90d2b145f4fb5ceaf53d112d905b65e9a5d6ee3141a917d39a21cfde"},"previousTransaction":"Fde5EQZvYY59Aj4ruZYvWdtWXUorsqHUJASosCdWAQvZ","storageRebate":0,"content":{"dataType":"moveObject","type":"0x2::staking_pool::StakedSui","hasPublicTransfer":false,"fields":{"delegation_activation_epoch":"0","id":{"id":"0x1ace65f54d65a96251b3f46bfa720ab65a7ebe0174caa9fa1dc0d65a56ae7872"},"pool_id":"0x628ffd0e51e9a6ea32c13c2739a31a8f344b557d3429e057b377a9c499b9bb13","principal":"25000000000000000","sui_token_lock":null,"validator_address":"0x00df836a90d2b145f4fb5ceaf53d112d905b65e9a5d6ee3141a917d39a21cfde"}},"bcs":{"dataType":"moveObject","type":"0x2::staking_pool::StakedSui","hasPublicTransfer":false,"version":1,"bcsBytes":"Gs5l9U1lqWJRs/Rr+nIKtlp+vgF0yqn6HcDWWlaueHJij/0OUemm6jLBPCc5oxqPNEtVfTQp4Fezd6nEmbm7EwDfg2qQ0rFF9Ptc6vU9ES2QW2XppdbuMUGpF9OaIc/eAAAAAAAAAAAAgGIXXtFYAAA="}}},"id":1}%  
```

### CODE REVIEW

File: crates/sui-indexer/src/models/objects.rs [hunk 1]
```diff
@@ -118,9 +133,161 @@ impl Object {
...
+                    u64::MAX,
+                )
+                .map_err(|e| {
+                    IndexerError::SerdeError(format!("Failed to parse move object: {}", e))
+                })
+            }?;
+
+            let owner = self.get_owner()?;
+            let obj = SuiBaseObject::new_move(move_obj.clone(), owner, previous_transaction);
+            let storage_rebate = obj.storage_rebate;
```
Comment (gegaowp) [inline crates/sui-indexer/src/models/objects.rs:195]: storage_rebate is always needed, which is derived from obj, as a result `owner`, `previous_transaction` and `move_obj` always need to be parsed, even though they are not rendered.

Comment (gegaowp) [reply crates/sui-indexer/src/models/objects.rs:195]: some optimizations can be done here for less parsing.

File: crates/sui-indexer/src/models/objects.rs [hunk 2]
```diff
@@ -118,9 +133,161 @@ impl Object {
...
+                    self.previous_transaction, e
+                ))
+            })?;
+
+            // only one BCS for non-package object
+            let obj_bcs = self
+                .bcs
+                .first()
+                .cloned()
+                .expect("Expect the BCS data to be non-empty")
```
Comment (gegaowp) [inline crates/sui-indexer/src/models/objects.rs:174]: throw err instead

File: crates/sui-indexer/src/apis/read_api.rs [hunk 3]
```diff
@@ -284,7 +301,7 @@ where
     ) -> RpcResult<Vec<SuiTransactionResponse>> {
         if self
             .method_to_be_forwarded
-            .contains(&"multi_get_transactions".to_string())
+            .contains(&"muti_get_transactions".to_string())
```
Comment (gegaowp) [inline crates/sui-indexer/src/apis/read_api.rs:304]: fix string.

**Review by 666lcz — APPROVED** [review]
Looks great!!

File: crates/sui-indexer/src/apis/read_api.rs [hunk 4]
```diff
@@ -212,6 +219,17 @@ where
             .await;
```
[the remark this answers is not in the document]
Comment (gegaowp) [reply crates/sui-indexer/src/apis/read_api.rs:219]: yep, did not see this method while drafting LOL

File: crates/sui-indexer/migrations/2022-12-01-034426_objects/up.sql [hunk 5]
```diff
@@ -23,6 +23,7 @@ CREATE TABLE objects
     previous_transaction   base58digest  NOT NULL,
     object_type            VARCHAR       NOT NULL,
     object_status          object_status NOT NULL,
+    has_public_transfer    BOOLEAN       NOT NULL,
```
[the remark this answers is not in the document]
Comment (patrickkuo) [reply crates/sui-indexer/migrations/2022-12-01-034426_objects/up.sql:26]: I will wait for this PR, my changes for get object is mainly to enable testing module cache

**Pull request merged**